File: 1.EarthEngineDownload/B01_B12.R
setwd("C:/Users/UTENTE/Google Drive laptop/LUCAS Copernicus/EarthEngine/Data")
# remotes::install_github("r-spatial/rgee")
library(rgee)
library(dplyr)
library(readr)
library(purrr)
library(tibble)
library(reticulate)

# ee_install_set_pyenv(
#   py_path = "C:\\Users\\UTENTE\\AppData\\Local\\Programs\\Python\\Python311",
#   py_env  = "rgee")
# ee_mod <- import("ee")
# versione_ee <- ee_mod$`__version__`
# cat("Versione Earth Engine Python API:", versione_ee, "\n")

# rgee::ee_check()
ee_clean_user_credentials()
# ee_Authenticate(auth_mode = "localhost")
# ee_Initialize(project = "landcovercopernicus", drive = FALSE, gcs = FALSE)
# Modalità “notebook”
ee_Initialize(project = "landcovercopernicus")
# Modalità “gcloud”, se hai già login tramite gcloud CLI
ee_Initialize(project = "landcovercopernicus", auth_mode = "gcloud")
# Parametri
input_csv <- "sample_points.csv"         # deve avere: POINT_ID, longitude, latitude
start_date <- "2018-04-01"
end_date   <- "2018-07-31"
batch_size <- 1000
output_csv <- "s2_bands_points.csv"

bands_needed <- c('B1', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8', 'B8A', 'B9', 'B11', 'B12')

# Leggi i punti e calcola la bounding box
points <- read_csv(input_csv, show_col_types = FALSE)
bbox <- st_bbox(
  st_as_sf(points, coords = c("longitude", "latitude"), crs = 4326)
)

ee_bbox <- ee$Geometry$Rectangle(
  coords = c(bbox["xmin"], bbox["ymin"], bbox["xmax"], bbox["ymax"])
)

# Prendi le immagini Sentinel-2 SR Harmonized nell'intervallo
collection <- ee$ImageCollection("COPERNICUS/S2_SR_HARMONIZED")$
  filterDate(start_date, end_date)$
  filterBounds(ee_bbox)$
  filter(ee$Filter$lt("CLOUDY_PIXEL_PERCENTAGE", 20))

# Crea il mosaico e seleziona solo le bande richieste
image <- collection$mosaic()$select(bands_needed)

# Funzione per processare un batch
process_batch <- function(batch_df) {
  # Crea una FeatureCollection dei punti con POINT_ID
  features <- map2(
    batch_df$longitude, batch_df$latitude,
    ~ ee$Feature(ee$Geometry$Point(c(.x, .y)), list(POINT_ID = .x))
  )
  fc <- ee$FeatureCollection(features)
  # Sample
  sampled <- tryCatch({
    image$sampleRegions(
      collection = fc,
      scale = 10,
      geometries = FALSE
    )$getInfo()
  }, error = function(e) NULL)
  if (is.null(sampled)) return(tibble())
  rows <- map_dfr(sampled$features, function(f) {
    as_tibble(f$properties)
  })
  return(rows)
}

# Processing in batch
n <- nrow(points)
batches <- split(points, ceiling(seq_along(1:n) / batch_size))

all_results <- list()
for (i in seq_along(batches)) {
  cat(sprintf("Processing batch %d of %d ...\n", i, length(batches)))
  batch_result <- process_batch(batches[[i]])
  all_results[[i]] <- batch_result
  # Salvataggio parziale
  bind_rows(all_results) %>% write_csv(output_csv, na = "")
  Sys.sleep(1)
}
final_df <- bind_rows(all_results)
write_csv(final_df, output_csv, na = "")
cat(sprintf("Completato! Output: %s (%d record)\n", output_csv, nrow(final_df)))

File: 1.EarthEngineDownload/B01_B12.py
import os
import ee
import pandas as pd
from tqdm import tqdm
import time

# os.chdir("D:\\Google Drive\\LUCAS Copernicus\\EarthEngine\\Data")
os.chdir("C:/Users/UTENTE/Google Drive laptop/LUCAS Copernicus/EarthEngine/Data")

# Initialize the Earth Engine library
ee.Authenticate(auth_mode='localhost')
ee.Initialize(project='landcovercopernicus')

# --- Carica i punti ---
start_date = '2018-04-01'
end_date   = '2018-07-31'
orbit_pass = 'ASCENDING'  # O 'DESCENDING', oppure togli per entrambe

batch_size = 1000
output_csv = 'B01_B12_Italy_2018.csv'

# --- Lista delle bande Sentinel-2 (SR Harmonized) ---
bands_needed = ['B1', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8', 'B8A', 'B9', 'B11', 'B12']

# --- Carica i punti e crea bounding box su tutta l'area ---
df = pd.read_csv("italy_points.csv")
bbox = ee.Geometry.BBox(df['longitude'].min(), df['latitude'].min(),
                        df['longitude'].max(), df['latitude'].max())

# --- Crea mosaico Sentinel-2 su tutta l'area ---
collection = ee.ImageCollection('COPERNICUS/S2_SR_HARMONIZED') \
    .filterDate(start_date, end_date) \
    .filterBounds(bbox) \
    .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', 20))

if collection.size().getInfo() == 0:
    raise Exception("Nessuna immagine Sentinel-2 trovata nel periodo e area selezionati!")

image = collection.mosaic().select(bands_needed)

# --- Funzione per processare un batch di punti ---
def process_batch(batch_df):
    features = []
    for _, row in batch_df.iterrows():
        point = ee.Geometry.Point([row['longitude'], row['latitude']])
        feature = ee.Feature(point, {'POINT_ID': int(row['POINT_ID'])})
        features.append(feature)
    fc = ee.FeatureCollection(features)
    try:
        sampled = image.sampleRegions(
            collection=fc,
            scale=10,
            geometries=False
        ).getInfo()
        rows = []
        for f in sampled['features']:
            props = f['properties']
            rows.append(props)
        return rows
    except Exception as e:
        print(f"Errore batch: {e}")
        return []

# --- Loop sui batch ---
all_results = []
num_batches = (len(df) + batch_size - 1) // batch_size

for i in tqdm(range(num_batches), desc="Batch processing"):
    start_idx = i * batch_size
    end_idx = min((i + 1) * batch_size, len(df))
    batch_df = df.iloc[start_idx:end_idx]
    batch_results = process_batch(batch_df)
    all_results.extend(batch_results)
    # Salva risultati parziali dopo ogni batch
    pd.DataFrame(all_results).to_csv(output_csv, index=False)
    time.sleep(1)  # evita troppi request ravvicinati

print(f"Completato! Output: {output_csv} ({len(all_results)} record)")
